### ISSUE TITLE
Make Junctions short-curcuit similar to how `and`/`or` do

### ISSUE THREAD

**Issue by zoffixznet** [rakudo/rakudo#1963]
This is really for The Future™, once our Junctions autothread; filing this so it doesn't get lost.

This is following on [the original discussion in the docs](https://github.com/perl6/doc/issues/2120#issuecomment-399996522) that Junctions were planned to short-circuit more than they do now, i.e. in the code below, one or more sub calls might not be executed at all, if one of the already-called subs returned a truthy value:

```perl6
    if a() | b() | c() {
        say "At least one of the routines was called and returned a truthy value"
    }
```

We probably should look at this after auto-threading is implemented, because we'll possibly will need to use the same mechanism we might use to figure out whether auto-threading the junction is worth it to decide whether thunking the args for short-circuiting is worth it too.

**Comment by lizmat** [comment]
My feeling is that we cannot guarantee anything in this respect: Junctions may be executed in parallel, or not.  If you really want short-circuiting behaviour, you should use || instead of | , because || is documented and guaranteed to short-circuit.

**Comment by zoffixznet** [comment]
> My feeling is that we cannot guarantee anything in this respect

Correct, there would be no order to it and we might execute more than needed.

The user-specific caution would be that they should not expect all sub calls in `a() | b() | c()` to actually be called, *not* that they can rely that only one of them would be called as long as it returns True.
